### ISSUE TITLE
Support human readable filtering in `google_folder` and/or `google_folders`

### ISSUE THREAD

**Issue by uplight-james** [hashicorp/terraform-provider-google#14076]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

Currently the [`google_folder`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/folder) data block allows you to get a folder only if you know its `id`, which is fed to the `folder` argument. 

This is fine if you happen to know that but its usefulness is limited as an end user is more likely to know a.) the name of the folder they want to target and b.) the name of its parent folder. It would be nice if given that information, you could look up your folders information. 

In addition, _most_ use cases for the `google_folder` are made obsolete by not supporting this functionality, as you usually would want to do this to actually discover the `id` of the folder to feed to other resources.

This also would work with googles [current folder naming convention](https://cloud.google.com/resource-manager/docs/creating-managing-folders):

> Folder display names must be unique within the same level of the hierarchy.

So this resource could reliably return exactly one resource.

### New or Affected Resource(s)

* [google_folder](https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/folder)
* [google_folders](https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/folders)

### Potential Terraform Configuration

**Option 1, support a `folder_name` and `parent_name` argument on `google_folder`**

```tf
data "google_folder" "first-folder" {
  folder_name  = "my_neat_folder"
  parent_name = "my_parent_folder_name"
}
```

**Option 2 (done in addition or instead of option 1), support a `filter` argument on [`google_folders`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/folders), similar to [`google_projects`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/projects)**

```tf
data "google_folder" "my_folders" {
  filter = "parent.name:some_parent_name folder.name:some_folder_name
}
```

**The problem with both of these solutions**

Currently, GCP does not support filtering on names like this (the closes thing seems to be `/v3/folders:search` but it only supports IDs) meaning that the terraform code itself would have to do a lot of listing/pagination to get all folders then their decedents.

I am also going to hit our solutions architect up with this to see if there is any way we could feature request filtering on the human readable "names" of folders.

### References

* #0000

**Comment by rileykarson** [comment]
Note: There is the https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/active_folder datasource, which allows searching for the display name of a folder but does require the id of the parent folder or organization. Not as great for nested folders, but should work for orgs or horizontal structures where top-level folders' ids are known.

One ambiguity I do want to call out is what "name" means- unfortunately while it meant the human-specified identifier in earlier GCP APIs, it was decided at some point that "name" should refer to the relative path of a resource per https://google.aip.dev/122. This matters a tiny bit here, where https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/active_folder#name and the display_name are distinct.

**Comment by uplight-james** [comment]
@rileykarson yes thank you for calling that out, I definitely mean `display_name` and will update the issue accordingly.

**Comment by rileykarson** [comment]
We could probably make `parent` conditional in `data.google_active_folder` when `:search` is the API method being used.
